src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
It's weakly pure, only changes stuff that's passed into the function and doesn't have a side-effect.
But for weakly pure functions any reference might alias global or shared data.
And also `shared` is a type qualifier not a storage class like `static`.

```d
void foo(ref shared(int) a) pure { a = 2; } // foo is weakly pure
shared int a;
void bar() { foo(a); } // bar is impure
```